CRM: remove built CSS from the monorepo, plus .gitignore and .gitattributes cleanup

coder-karen opened the pull request:
## Proposed changes:

* This PR removes all built CRM CSS in the monorepo (CSS that is built via Webpack from SCSS files). The CSS files will still be built and added to the mirror repo (for production).
* The PR also updates the `.gitattributes` and `.gitignore` files, as well as removing the `.distignore` file which serves no obvious purpose now that CRM is built and deployed using monorepo tooling.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

https://github.com/Automattic/zero-bs-crm/issues/3056

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

* To test this PR out, check out this branch in a local development environment.
* Run `jetpack build plugins/crm` - by doing so this should not add new untracked files, or files with changes needing committing.
* Run `jetpack build plugins/crm --production` - the results should be the same.
* Try adding a change to a SCSS file and run a build again. You should see the change in the CSS file locally, but the file won't show as a changed file needing commit (it will continue to show as a deleted file with this PR checked out).

coder-karen committed:
Fixing merge conflict - removing additional built CSS files

coder-karen committed:
[not verified] changelog

coder-karen committed:
[not verified] Removing export-ignore entries from .gitttributes

tbradsha commented on projects/plugins/crm/.gitattributes:
```diff
@@ -1,24 +1,7 @@
-codeception.yml      export-ignore
-codeception.dist.yml export-ignore
-.wp-env.json         export-ignore
-/.wordpress-org/**/* export-ignore
-/.wordpress-org      export-ignore
-.wordpress-org/**/*  export-ignore
-.wordpress-org       export-ignore
```
I assume this won't impact the ability for this folder to control the WP.org repo assets? Or if it does, how are said assets updated going forward?

tbradsha replied:
Also, any idea what the different between `export-ignore` and `production-exclude` are?

coder-karen replied:
Looking into it, it (`export-ignore`) is related to packages (I don't see a Jetpack CRM Packagist entry for CRM), and the `export-ignore` meant for building an archive.

I am not too sure how CRM use `export-ignore` previously in `.gitattributes` - that's all there were in there. Likely in the way the plugin was built / packaged. Now we're using `production-include` and `production-exclude`. 

As for `production-exclude`, and `production-include`, they're focussed on what is being built based on what is already in the monorepo. For example `.gitignore` makes sure we don't commit changes to the 'build' directory, but when we're creating a release and the release is added to the mirror repo, in the process we're running the production build commands and the build directory is built. We do want that in production (just not in the monorepo), so we `production-include` it. Same for `production-exclude`, but it's for things we want in the monorepo and not in the mirror repo / production.

I've also been using the other standalones and their `.gitattributes` as guides ([Social](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/social/.gitattributes), [Search](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/search/.gitattributes) for example). For example we previously included node-modules as an export-ignore, however this is now not included in `.gitattributes` (just in .`gitignore`), also in line with the other standalones.

As for `.wordpress-org`, in 5.6.0 that was in both the mirror repo and the download on SVN. Checking 5.5.3 I see it wasn't available in the download, however I'm also not too sure how that is used to build the assets so  I'm not too sure about adding that to `production-exclude`. It was in the `.distignore` for 5.6.0 as well but still displays in the plugin download.

tbradsha commented on projects/plugins/crm/.gitattributes:
```diff
@@ -1,24 +1,7 @@
 # Files to include in the mirror repo, but excluded via gitignore
 # Remember to end all directories with `/**` to properly tag every file.
 # /src/js/example.min.js  production-include
+build/**                                        production-include
```
Is this something we use at the moment?

coder-karen replied:
It will be. It was added as part of this PR - https://github.com/Automattic/jetpack/pull/29062 - and will be included in local development and production builds (but not the monorepo). It just looks like it was never added to `production-include` in `gitattributes` at the time.

tbradsha commented on projects/plugins/crm/.gitattributes:
```diff
@@ -29,11 +12,7 @@ vendor/dompdf/**                                production-include
-**/css/*.min.css                                production-include
-**/css/welcome-to-zbs/*.min.css                 production-include
-**/css/integrations/*                           production-include
+**/css/**                                       production-include
```
Yay condensing!

tbradsha commented on projects/plugins/crm/.gitignore:
```diff
@@ -1,37 +1,20 @@
 templates/invoices/index.html
 templates/quotes/.htaccess
-templates/quotes/index.html
-tests/acceptance.suite.yml
```
I'm not sure we want to remove the first one; that's used for running acceptance tests with a local config.

coder-karen replied:
Ah thanks! I did not see it in my local environment so assumed it was outdated. I'll add that back.

tbradsha commented on projects/plugins/crm/.gitattributes:
```diff
@@ -1,24 +1,7 @@
 # Remember to end all directories with `/**` to properly tag every file.
 # /src/js/example.min.js  production-include
+build/**                                        production-include
 jetpack_vendor/**                               production-include
```
Is this one we need in production?

coder-karen replied:
It is: https://github.com/Automattic/jetpack/pull/30058#discussion_r1166544608

tbradsha commented on projects/plugins/crm/.gitignore:
```diff
@@ -1,37 +1,20 @@
 **/js/welcome-to-zbs/*.min.js
 **/js/*.min.js.LICENSE.txt
-.cache/
-phpcs.xml
-phpcs.security.xml
```
These are overrides for the distribution `.xml`, so we should probably leave them (or remove PHPCS config from the repo altogether).

coder-karen replied:
Not sure I follow. I couldn't find the files locally - when you say these are overrides does that mean these might be generated in some situations?

coder-karen replied:
Ah just saw your follow-on comment below, will follow up in another comment.

tbradsha reviewed (commented):
I started to review this, but got a bit confused as to what needed to be where. I definitely have some reservations about the `.gitignore`.

As an aside, it seems there's likely some redundancy and inconsistency with the wildcards and slashes.

tbradsha commented:
I switched to the branch, and immediately got these as untracked:

```
phpcs.security.xml
tests/acceptance.suite.yml
```

It looks like some other files are ignored by the `.gitignore` in the `jetpack` repo root (e.g. `.idea` and `phpcs.xml`).

>Run `jetpack build plugins/crm` - by doing so this should not add new untracked files, or files with changes needing committing.

✅ 

> Run` jetpack build plugins/crm --production` - the results should be the same

✅ 

> Try adding a change to a SCSS file and run a build again

✅ Confirmed that the SCSS shows as changed but the CSS files do not show as tracked.

---

All this looks good, but I'm a bit confused as to how to test the actual production build. 🤔

tbradsha commented:
Downloaded the beta zip from live branches, and I still see the following that may or may not be wrong:

```
.gitattributes		# we should remove this
.wordpress-org		# do we need this?
build			# do we need this?
composer.json		# we should remove this
docs			# this appears to be a beta artifact
jetpack_vendor		# do we need this?
SECURITY.md		# this appears to belong to JP builds, as it references the JP plugin
version.txt		# do we need this?
```

I'm not sure how much of that is related to the beta build vs. final build.

coder-karen commented:
Thanks @tbradsha for reviewing this!

> As an aside, it seems there's likely some redundancy and inconsistency with the wildcards and slashes.

As for `.gitattributes`, I see the lack of consistency with initial directory slashes (that's what I get for initially copying from different standalones like [Social](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/social/.gitattributes) and [Jetpack](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/.gitattributes)), so I've fixed that up now. The glob usage is needed (in both `.gitattributes` and `.gitignore`) because JS and CSS directories aren't necessarily at root level. Is that part of what you were referring to? It seems `.gitignore` follows a different pattern than `.gitattributes` so I removed the `*` from the end of `**/css/integrations/`, that file is now consistent (commit to follow shortly).

> I switched to the branch, and immediately got these as untracked:
> phpcs.security.xml
> tests/acceptance.suite.yml

Interesting, I don't have those locally, but better safe than sorry so I'll add them back in.

Will follow up on your second comment with an additional comment here as well.

coder-karen commented:
> All this looks good, but I'm a bit confused as to how to test the actual production build. 🤔

Apart from `.gitignore` I'm not too sure - the beta zip looks like a partial option like you discovered. Other than that, comparing with other standalones has been my approach (what do they have included or not included, what is showing in production).

> Downloaded the beta zip from live branches, and I still see the following that may or may not be wrong:

```
.gitattributes		# we should remove this
.wordpress-org		# do we need this?
build			# do we need this?
composer.json		# we should remove this
docs			# this appears to be a beta artifact
jetpack_vendor		# do we need this?
SECURITY.md		# this appears to belong to JP builds, as it references the JP plugin
version.txt		# do we need this?
```

`.gitattributes` is excluded from the [mirror repo](https://github.com/Automattic/jetpack-crm) but it still visible there - but it isn't a complete file. It isn't showing in the 5.6.0 WP org download though.

`.wordpress-org` - we don't need that in SVN (in trunk), but Jetpack must follow a different approach in terms of adding assets (I think they are actually added manually). So I think it isn't ideal having that folder there but it seems we'd need to add the assets manually, separately, otherwise. I wouldn't remove it unless we've looked into how best to add the assets otherwise.

`build` - yes - see https://github.com/Automattic/jetpack/pull/30058#discussion_r1166544608. 

`composer.json` - that's visible in downloads of the Jetpack plugin and other standalone plugins. Keeping it appears to be [by intention](https://github.com/Automattic/jetpack/blob/trunk/docs/monorepo.md#creating-a-new-plugin) (though that may be intended for consumption of the plugin as a package, which is the case with some but not all standalones - though all have the composer.json file visible). That said the composer.json in the mirror repo is needed for things like the autotagger action. As far as I'm aware, if it's in the mirror repo it would then be in SVN.

`docs` - yes that's not showing in the mirror repo or most recent WP org download and I haven't made any changes to that in this PR.

`jetpack_vendor` - yes that a `production-include`, included in Jetpack and standalones, just like `vendor` (with all package info).

`SECURITY.md` - that's a result of the build process, the same file also included in all standalones.

`version.txt` - it seems that's beta related, it's not something I see in the mirror repo or production.

coder-karen committed:
[not verified] Some cleanup in .gitignore and .gitattributes based on PR review

tbradsha commented on projects/plugins/crm/.gitignore:
```diff
@@ -1,37 +1,22 @@
+/templates/invoices/index.html
+/templates/quotes/.htaccess
+/templates/quotes/index.html
+/templates/acceptance.suite.yml
```
```suggestion
/tests/acceptance.suite.yml
```

coder-karen replied:
Thanks!

tbradsha reviewed (changes requested):
One typo, but otherwise I think at least this is an iteration toward a cleaner setup.

As for the beta build `docs` and `SECURITY.md`, any idea the best way to get those cleaned up?

coder-karen commented:
> One typo, but otherwise I think at least this is an iteration toward a cleaner setup.

Thanks for catching that and re-reviewing!

> As for the beta build docs and SECURITY.md, any idea the best way to get those cleaned up?

The beta build docs are included in every beta build it seems, with the testing instructions doc. Since it's only shipping in the beta zips it doesn't seem super relevant to work out how (and if) we should be cleaning that up for our own releases.

I think I may be misunderstanding you though - by cleaning up are you referring to actually changing the content in them? Looks like the testing docs are copied [from this line](https://github.com/Automattic/jetpack/blob/trunk/.github/workflows/build.yml#L200) in build.yml, and the SECURITY.md file [copied over here](https://github.com/Automattic/jetpack/blob/trunk/tools/cli/commands/build.js#L731), in build.js. So changing the content or removing them would likely mean adding to those actions / scripts, so doesn't seem relevant to this PR anyway 😄

coder-karen committed:
Update projects/plugins/crm/.gitignore

Co-authored-by: tbradsha <32492176+tbradsha@users.noreply.github.com>

tbradsha commented:
> So changing the content or removing them would likely mean adding to those actions / scripts, so doesn't seem relevant to this PR anyway 😄

Agreed! Mainly I want to report it to the correct team. I'll put in two separate issues (Edit: #30102 and #30104).
